salt/_states/caasp_cmd.py

```diff
@@ -0,0 +1,202 @@
+        If the command has not terminated after timeout seconds, send the
+        subprocess sigterm, and if sigterm is ignored, follow up with sigkill
+
+    ignore_timeout
+        Ignore the timeout of commands, which is useful for running nohup
+        processes.
+
+        .. versionadded:: 2015.8.0
+
+    creates
+        Only run if the file or files specified by ``creates`` do not exist.
+
+        .. versionadded:: 2014.7.0
+
+    use_vt
+        Use VT utils (saltstack) to stream the command output more
+        interactively to the console and the logs.
+        This is experimental.
+
+    bg
+        If ``True``, run command in background and do not await or deliver it's
+        results.
+
+        .. versionadded:: 2016.3.6
+
+    .. note::
+
+        caasp_cmd.run supports the usage of ``reload_modules``. This functionality
+        allows you to force Salt to reload all modules. You should only use
+        ``reload_modules`` if your caasp_cmd.run does some sort of installation
+        (such as ``pip``), if you do not reload the modules future items in
+        your state which rely on the software being installed will fail.
+
+        .. code-block:: yaml
+
+            getpip:
+              caasp_cmd.run:
+                - name: /usr/bin/python /usr/local/sbin/get-pip.py
+                - unless: which pip
+                - require:
+                  - pkg: python
+                  - file: /usr/local/sbin/get-pip.py
+                - reload_modules: True
+
+    retry
+        This allows you to provide `attempts` and `interval`, what will retry the command as much
+        ``attempts`` times, separated by `interval` seconds.
+
+    '''
+    ### NOTE: The keyword arguments in **kwargs are passed directly to the
+    ###       ``cmd.run_all`` function and cannot be removed from the function
+    ###       definition, otherwise the use of unsupported arguments in a
+    ###       ``caasp_cmd.run`` state will result in a traceback.
+
+    retry_ = {'attempts': 1,
+              'interval': 1}
+    retry_.update(retry)
+
+    ret = None
+
+    for attempt in xrange(retry_['attempts']):
+        ret = __states__['cmd.run'](name=name,
+                                    onlyif=onlyif,
+                                    unless=unless,
+                                    creates=creates,
+                                    cwd=cwd,
+                                    runas=runas,
+                                    shell=shell,
+                                    env=env,
+                                    stateful=stateful,
+                                    umask=umask,
+                                    output_loglevel=output_loglevel,
+                                    quiet=quiet,
+                                    timeout=timeout,
+                                    ignore_timeout=ignore_timeout,
+                                    use_vt=use_vt,
+                                    **kwargs)
```

Comment:
Instead of doing:

```python
def run(name,
        onlyif=None,
        unless=None,
        creates=None,
        cwd=None,
        runas=None,
        shell=None,
        env=None,
        stateful=False,
        umask=None,
        output_loglevel='debug',
        quiet=False,
        timeout=None,
        ignore_timeout=False,
        use_vt=False,
        retry={},
        **kwargs):

    # ...

    ret = __states__['cmd.run'](name=name,
                                    onlyif=onlyif,
                                    unless=unless,
                                    creates=creates,
                                    cwd=cwd,
                                    runas=runas,
                                    shell=shell,
                                    env=env,
                                    stateful=stateful,
                                    umask=umask,
                                    output_loglevel=output_loglevel,
                                    quiet=quiet,
                                    timeout=timeout,
                                    ignore_timeout=ignore_timeout,
                                    use_vt=use_vt,
                                    **kwargs)
```

couldn't we just do?

```python
def run(name, **kwargs):
    # ...
    ret = __states__['cmd.run'](name, **kwargs)
```

Do we need to force default values for those arguments?

Reply:
I think that's needed because @ereslibre introduced the new `retry` dictionary option. However it would be possible to change the code to take a `kwargs`, remove the `retry` option and hand it over to `cmd.run`...

Reply:
Yes, it will be cleaner. Will refactor it, thanks for your feedback @inercia and @flavio.